File: tracks/haskell/exercises/variables-and-types/standard/starter/ConfigTest.hs
module Main where

-- Run with: runhaskell ConfigTest.hs

import Config

-- Simple test runner (no dependencies needed)
assert :: String -> Bool -> IO ()
assert label True  = putStrLn ("  PASS: " ++ label)
assert label False = putStrLn ("  FAIL: " ++ label)

assertLeft :: String -> Either String a -> IO ()
assertLeft label (Left _)  = putStrLn ("  PASS: " ++ label)
assertLeft label (Right _) = putStrLn ("  FAIL: " ++ label ++ " (expected Left, got Right)")

assertRight :: Show a => String -> (a -> Bool) -> Either String a -> IO ()
assertRight label _ (Left e)    = putStrLn ("  FAIL: " ++ label ++ " (got Left: " ++ e ++ ")")
assertRight label check (Right a)
  | check a   = putStrLn ("  PASS: " ++ label)
  | otherwise  = putStrLn ("  FAIL: " ++ label ++ " (got Right: " ++ show a ++ ")")

main :: IO ()
main = do
  putStrLn "=== Config Parser Tests ===\n"

  -- Test 1: Full environment
  putStrLn "Test: Full environment"
  let fullEnv = [ ("HOST", "api.example.com")
                , ("PORT", "9090")
                , ("DEBUG", "true")
                , ("MAX_RETRIES", "5")
                , ("TIMEOUT_SEC", "30.5")
                , ("ENVIRONMENT", "production")
                ]
  assertRight "parses successfully" (const True) (parseConfig fullEnv)
  assertRight "host is correct" (\c -> cfgHost c == "api.example.com") (parseConfig fullEnv)
  assertRight "port is 9090" (\c -> cfgPort c == Port 9090) (parseConfig fullEnv)
  assertRight "debug is True" (\c -> cfgDebug c == True) (parseConfig fullEnv)
  assertRight "maxRetries is Just 5" (\c -> cfgMaxRetries c == Just 5) (parseConfig fullEnv)
  assertRight "timeout is 30.5" (\c -> cfgTimeout c == TimeoutSec 30.5) (parseConfig fullEnv)
  assertRight "env is production" (\c -> cfgEnv c == "production") (parseConfig fullEnv)

  -- Test 2: Defaults applied
  putStrLn "\nTest: Defaults applied (only HOST)"
  let minEnv = [("HOST", "localhost")]
  assertRight "parses successfully" (const True) (parseConfig minEnv)
  assertRight "host is localhost" (\c -> cfgHost c == "localhost") (parseConfig minEnv)
  assertRight "port defaults to 8080" (\c -> cfgPort c == Port 8080) (parseConfig minEnv)
  assertRight "debug defaults to False" (\c -> cfgDebug c == False) (parseConfig minEnv)
  assertRight "maxRetries is Nothing" (\c -> cfgMaxRetries c == Nothing) (parseConfig minEnv)
  assertRight "timeout defaults to 30.0" (\c -> cfgTimeout c == TimeoutSec 30.0) (parseConfig minEnv)
  assertRight "env defaults to development" (\c -> cfgEnv c == "development") (parseConfig minEnv)

  -- Test 3: Missing HOST
  putStrLn "\nTest: Missing HOST"
  assertLeft "returns Left" (parseConfig [("PORT", "8080")])

  -- Test 4: Empty HOST
  putStrLn "\nTest: Empty HOST"
  assertLeft "returns Left" (parseConfig [("HOST", "")])

  -- Test 5: Invalid PORT
  putStrLn "\nTest: Invalid PORT"
  assertLeft "returns Left" (parseConfig [("HOST", "localhost"), ("PORT", "abc")])

  -- Test 6: MAX_RETRIES = 0 (intentionally zero, not missing)
  putStrLn "\nTest: MAX_RETRIES explicitly set to 0"
  let zeroRetries = [("HOST", "localhost"), ("MAX_RETRIES", "0")]
  assertRight "maxRetries is Just 0, not Nothing"
    (\c -> cfgMaxRetries c == Just 0) (parseConfig zeroRetries)

  -- Test 7: Debug variants
  putStrLn "\nTest: DEBUG variants"
  let mkEnv d = [("HOST", "localhost"), ("DEBUG", d)]
  assertRight "\"true\" -> True" (\c -> cfgDebug c == True) (parseConfig (mkEnv "true"))
  assertRight "\"false\" -> False" (\c -> cfgDebug c == False) (parseConfig (mkEnv "false"))
  assertRight "\"1\" -> True" (\c -> cfgDebug c == True) (parseConfig (mkEnv "1"))
  assertRight "\"0\" -> False" (\c -> cfgDebug c == False) (parseConfig (mkEnv "0"))

  -- Test 8: Invalid TIMEOUT_SEC
  putStrLn "\nTest: Invalid TIMEOUT_SEC"
  assertLeft "returns Left" (parseConfig [("HOST", "localhost"), ("TIMEOUT_SEC", "abc")])

  -- Test 9: showConfig doesn't crash
  putStrLn "\nTest: showConfig"
  case parseConfig minEnv of
    Left e  -> putStrLn ("  FAIL: parse failed: " ++ e)
    Right c -> do
      let s = showConfig c
      assert "returns non-empty string" (not (null s))
      putStrLn ("  Output: " ++ s)

  putStrLn "\n=== Done ==="

File: tracks/go/exercises/variables-and-types/standard/starter/config_test.go
package config

import "testing"

// Run with: go test -v ./...

func TestLoadConfig_FullEnv(t *testing.T) {
	// All fields provided — nothing should use defaults
	env := map[string]string{
		"HOST":         "api.example.com",
		"PORT":         "9090",
		"DEBUG":        "true",
		"MAX_RETRIES":  "5",
		"TIMEOUT_SEC":  "30.5",
		"ENVIRONMENT":  "production",
	}

	cfg, err := LoadConfig(env)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if cfg.Host != "api.example.com" {
		t.Errorf("Host = %q, want %q", cfg.Host, "api.example.com")
	}
	if cfg.Port != 9090 {
		t.Errorf("Port = %d, want %d", cfg.Port, 9090)
	}
	if cfg.Debug != true {
		t.Errorf("Debug = %v, want %v", cfg.Debug, true)
	}
	if cfg.MaxRetries == nil {
		t.Fatal("MaxRetries is nil, want *5")
	}
	if *cfg.MaxRetries != 5 {
		t.Errorf("MaxRetries = %d, want %d", *cfg.MaxRetries, 5)
	}
	if cfg.TimeoutSec != 30.5 {
		t.Errorf("TimeoutSec = %f, want %f", cfg.TimeoutSec, 30.5)
	}
	if cfg.Environment != "production" {
		t.Errorf("Environment = %q, want %q", cfg.Environment, "production")
	}
}

func TestLoadConfig_DefaultsApplied(t *testing.T) {
	// Only required field provided — everything else should use defaults
	env := map[string]string{
		"HOST": "localhost",
	}

	cfg, err := LoadConfig(env)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if cfg.Host != "localhost" {
		t.Errorf("Host = %q, want %q", cfg.Host, "localhost")
	}
	if cfg.Port != DefaultPort {
		t.Errorf("Port = %d, want default %d", cfg.Port, DefaultPort)
	}
	if cfg.Debug != false {
		t.Errorf("Debug = %v, want default false", cfg.Debug)
	}
	if cfg.MaxRetries != nil {
		t.Errorf("MaxRetries = %v, want nil (not set)", cfg.MaxRetries)
	}
	if cfg.TimeoutSec != DefaultTimeoutSec {
		t.Errorf("TimeoutSec = %f, want default %f", cfg.TimeoutSec, DefaultTimeoutSec)
	}
	if cfg.Environment != DefaultEnvironment {
		t.Errorf("Environment = %q, want default %q", cfg.Environment, DefaultEnvironment)
	}
}

func TestLoadConfig_MissingHost(t *testing.T) {
	// Host is required — should error
	env := map[string]string{
		"PORT": "8080",
	}

	_, err := LoadConfig(env)
	if err == nil {
		t.Fatal("expected error for missing HOST, got nil")
	}
}

func TestLoadConfig_EmptyHost(t *testing.T) {
	// Host present but empty — should error
	env := map[string]string{
		"HOST": "",
	}

	_, err := LoadConfig(env)
	if err == nil {
		t.Fatal("expected error for empty HOST, got nil")
	}
}

func TestLoadConfig_InvalidPort(t *testing.T) {
	env := map[string]string{
		"HOST": "localhost",
		"PORT": "not-a-number",
	}

	_, err := LoadConfig(env)
	if err == nil {
		t.Fatal("expected error for invalid PORT, got nil")
	}
}

func TestLoadConfig_MaxRetriesZero(t *testing.T) {
	// MAX_RETRIES=0 is intentional — should be *int pointing to 0, not nil
	env := map[string]string{
		"HOST":        "localhost",
		"MAX_RETRIES": "0",
	}

	cfg, err := LoadConfig(env)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	if cfg.MaxRetries == nil {
		t.Fatal("MaxRetries is nil, want *0 (explicitly set to zero)")
	}
	if *cfg.MaxRetries != 0 {
		t.Errorf("MaxRetries = %d, want 0", *cfg.MaxRetries)
	}
}

func TestLoadConfig_DebugVariants(t *testing.T) {
	tests := []struct {
		input string
		want  bool
	}{
		{"true", true},
		{"false", false},
		{"1", true},
		{"0", false},
	}

	for _, tt := range tests {
		env := map[string]string{
			"HOST":  "localhost",
			"DEBUG": tt.input,
		}

		cfg, err := LoadConfig(env)
		if err != nil {
			t.Errorf("DEBUG=%q: unexpected error: %v", tt.input, err)
			continue
		}
		if cfg.Debug != tt.want {
			t.Errorf("DEBUG=%q: got %v, want %v", tt.input, cfg.Debug, tt.want)
		}
	}
}

func TestLoadConfig_InvalidTimeout(t *testing.T) {
	env := map[string]string{
		"HOST":        "localhost",
		"TIMEOUT_SEC": "abc",
	}

	_, err := LoadConfig(env)
	if err == nil {
		t.Fatal("expected error for invalid TIMEOUT_SEC, got nil")
	}
}

func TestConfig_String(t *testing.T) {
	env := map[string]string{
		"HOST": "localhost",
	}

	cfg, err := LoadConfig(env)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}

	s := cfg.String()
	if s == "" {
		t.Error("String() returned empty string")
	}
	// Just verify it doesn't panic and returns something non-empty.
	// Exact format is up to the implementer.
	t.Logf("Config.String() output:\n%s", s)
}
